src/python/gudhi/wasserstein.py

```diff
@@ -62,36 +62,79 @@ def _perstot(X, order, internal_p):
     return (np.sum(np.linalg.norm(X - Xdiag, ord=internal_p, axis=1)**order))**(1./order)
 
 
-def wasserstein_distance(X, Y, order=2., internal_p=2.):
+def _clean_match(match, n, m):
     '''
-    :param X: (n x 2) numpy.array encoding the (finite points of the) first diagram. Must not contain essential points (i.e. with infinite coordinate).
+    :param match: a list of the form [(i,j) ...]
+    :param n: int, size of the first dgm
+    :param m: int, size of the second dgm
+    :return: a modified version of match where indices greater than n, m are replaced by -1, encoding the diagonal.
+                and (-1, -1) are removed
+    '''
+    new_match = []
+    for i,j in match:
+        if i >= n:
+            if j < m:
+               new_match.append((-1, j))
```

Comment:
Can't you do something like this?
```
P[-1,-1]=0
new=match.nonzero()
new[0][new[0]>=n]=-1
new[1][new[1]>=m]=-1
```
(or some variant to give whatever output we decide is the right one)

Reply:
Indeed it seems much more compact :) It changes the output type, but I have no reason to prefer one over the other.

Reply:
If we output everything at once, with -1 to represent the diagonal, then either a numpy array (n,2) or a pair of arrays would be fine (possibly some other variants as well). It could also be convenient to return an array (n,2) for regular pairs, and 2 1D arrays for matchings with the diagonal. I can get those from the combined array using `np.compress(a[:,0]==-1,a[:,1])` or similar, but if most users are going to do this anyway, we could do it for them (if it is easier for us than it is for them). Then the question is back to what users want it for...

Reply:
The advantage of returning a (n x 2) np.array is that it is rather compact ; also as you mentioned earlier using `-1` to represent the diagonal is convenient as, say, if one has `(- 1 , j )` in the matching, the corresponding cost can actually be accessed by `C[-1, j]`. 


(It makes me realize that we do not provide a way for the user to build the distance matrix, although it might be useful... Should we do that? It would just consist in removing the `_` in `_build_dist_matrix` I think.)

Reply:
> The advantage of returning a (n x 2) np.array is that it is rather compact

Yes. The drawback is if "most" users split it in 3 afterwards... Of course there is the possibility of having an option to choose which type of output we want. That could be overkill.

> also as you mentioned earlier using `-1` to represent the diagonal is convenient as, say, if one has `(- 1 , j )` in the matching, the corresponding cost can actually be accessed by `C[-1, j]`.

Keeping `n` and `m` also works for that, although it is less convenient for other purposes.

> (It makes me realize that we do not provide a way for the user to build the distance matrix, although it might be useful... Should we do that? It would just consist in removing the `_` in `_build_dist_matrix` I think.)

What use do you see for that distance matrix? If I use the distance matrix, it will be one built with pytorch, jax, or some other autodifferentiable type, so I won't be able to use your function that explicitly uses scipy.spatial.distance.cdist. But maybe there are other reasons to want that matrix?

Reply:
Well it is hard for me to anticipate what people would do with the matching... I think the compact (n x 2) output is the way to go. 

About returning the matrix, I don't have much idea. It is just cheap to provide this option to the user, although it make the doc heavier. Let's forget that for now.

Reply:
> Well it is hard for me to anticipate what people would do with the matching... I think the compact (n x 2) output is the way to go.

ok for me.